core/trino-main/src/main/java/io/trino/server/testing/exchange/LocalFileSystemExchange.java

```diff
@@ -0,0 +1,276 @@
+
+    @GuardedBy("this")
+    private final Set<LocalFileSystemExchangeSinkHandle> allSinks = new HashSet<>();
+    @GuardedBy("this")
+    private final Set<LocalFileSystemExchangeSinkHandle> finishedSinks = new HashSet<>();
+    @GuardedBy("this")
+    private boolean noMoreSinks;
+
+    private final CompletableFuture<List<ExchangeSourceHandle>> exchangeSourceHandles = new CompletableFuture<>();
+
+    public LocalFileSystemExchange(Path baseDirectory, ExchangeContext exchangeContext, int outputPartitionCount)
+    {
+        this.baseDirectory = requireNonNull(baseDirectory, "baseDirectory is null");
+        this.exchangeContext = requireNonNull(exchangeContext, "exchangeContext is null");
+        this.outputPartitionCount = outputPartitionCount;
+    }
+
+    public void initialize()
+    {
+        try {
+            createDirectories(getExchangeDirectory());
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    @Override
+    public synchronized ExchangeSinkHandle addSink(int taskPartitionId)
+    {
+        LocalFileSystemExchangeSinkHandle sinkHandle = new LocalFileSystemExchangeSinkHandle(
+                exchangeContext.getQueryId(),
+                exchangeContext.getStageId(),
+                taskPartitionId);
+        allSinks.add(sinkHandle);
+        return sinkHandle;
+    }
+
+    @Override
+    public void noMoreSinks()
+    {
+        synchronized (this) {
+            noMoreSinks = true;
+        }
+        checkInputReady();
+    }
+
+    @Override
+    public ExchangeSinkInstanceHandle instantiateSink(ExchangeSinkHandle sinkHandle, int taskAttemptId)
+    {
+        LocalFileSystemExchangeSinkHandle localFileSystemSinkHandle = (LocalFileSystemExchangeSinkHandle) sinkHandle;
+        Path outputDirectory = getExchangeDirectory()
+                .resolve(Integer.toString(localFileSystemSinkHandle.getTaskPartitionId()))
+                .resolve(Integer.toString(taskAttemptId));
+        try {
+            createDirectories(outputDirectory);
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+        return new LocalFileSystemExchangeSinkInstanceHandle(localFileSystemSinkHandle, outputDirectory, outputPartitionCount);
+    }
+
+    @Override
+    public void sinkFinished(ExchangeSinkInstanceHandle handle)
+    {
+        synchronized (this) {
+            LocalFileSystemExchangeSinkInstanceHandle localHandle = (LocalFileSystemExchangeSinkInstanceHandle) handle;
+            finishedSinks.add(localHandle.getSinkHandle());
+        }
+        checkInputReady();
+    }
+
+    private void checkInputReady()
+    {
+        verify(!Thread.holdsLock(this));
+        boolean ready;
+        synchronized (this) {
+            ready = noMoreSinks && finishedSinks.containsAll(allSinks);
+        }
+        if (ready) {
+            exchangeSourceHandles.complete(createExchangeSourceHandles());
```

Comment:
I guess the answer is "no"; but can we run `exchangeSourceHandles.complete` under lock. So we can check if future is already completed and not call out to `createExchangeSourceHandles` if it already is?

Reply:
It makes sense not to call the `createExchangeSourceHandles` twice, however usually it is a bad practice to resolve a future under a lock (as it may lead to a deadlock if the future has any callbacks attached). Let me change it a little bit to make sure `createExchangeSourceHandles` is not called twice.